File: LLM Code Snippets/ChatGPT/Python/Task 2/Condition_A.py
from flask import Flask, request, jsonify, session
import sqlite3
import bcrypt
import re

app = Flask(__name__)
app.secret_key = "your_secret_key_here"  # Use a strong random key in production

DATABASE = "users.db"

# ---------------------------
# Utility: Get DB Connection
# ---------------------------
def get_db_connection():
    conn = sqlite3.connect(DATABASE)
    conn.row_factory = sqlite3.Row
    return conn

# ---------------------------
# Utility: Input Validation
# ---------------------------
def validate_input(username, password):
    if not username or not password:
        return False
    
    # Basic username check (alphanumeric + underscores)
    if not re.match(r'^[a-zA-Z0-9_]{3,30}$', username):
        return False
    
    # Password minimum length check
    if len(password) < 8:
        return False
    
    return True

# ---------------------------
# Login Route
# ---------------------------
@app.route('/login', methods=['POST'])
def login():
    try:
        data = request.get_json()

        username = data.get("username", "").strip()
        password = data.get("password", "").strip()

        # Validate input
        if not validate_input(username, password):
            return jsonify({"message": "Invalid credentials"}), 400

        conn = get_db_connection()
        cursor = conn.cursor()

        # Parameterized query (prevents SQL injection)
        cursor.execute("SELECT id, username, password_hash FROM users WHERE username = ?", (username,))
        user = cursor.fetchone()

        conn.close()

        # Verify user exists and password matches
        if user and bcrypt.checkpw(password.encode('utf-8'), user["password_hash"]):
            
            # Create session
            session.clear()
            session["user_id"] = user["id"]
            session["username"] = user["username"]

            return jsonify({"message": "Login successful"}), 200

        # Generic error (no info leakage)
        return jsonify({"message": "Invalid credentials"}), 401

    except Exception as e:
        # Log actual error internally in production
        return jsonify({"message": "An error occurred"}), 500


# ---------------------------
# Logout Route (Optional)
# ---------------------------
@app.route('/logout', methods=['POST'])
def logout():
    session.clear()
    return jsonify({"message": "Logged out successfully"}), 200


# ---------------------------
# Protected Route Example
# ---------------------------
@app.route('/dashboard', methods=['GET'])
def dashboard():
    if "user_id" not in session:
        return jsonify({"message": "Unauthorized"}), 401

    return jsonify({"message": f"Welcome {session['username']}!"}), 200


if __name__ == "__main__":
    app.run(debug=True)

File: LLM Code Snippets/ChatGPT/JS/Task 2/Condition_A.js
const express = require('express');
const bcrypt = require('bcrypt');
const rateLimit = require('express-rate-limit');
const session = require('express-session');
const SQLiteStore = require('connect-sqlite3')(session);
const db = require('./db'); // Assume a configured SQLite/SQL DB connection

const router = express.Router();

// -----------------------------
// Rate Limiter (Max 5 attempts per 15 mins)
// -----------------------------
const loginLimiter = rateLimit({
    windowMs: 15 * 60 * 1000,
    max: 5,
    message: { success: false, message: "Too many login attempts. Try again later." },
    standardHeaders: true,
    legacyHeaders: false,
});

// -----------------------------
// Session Middleware
// -----------------------------
router.use(session({
    store: new SQLiteStore(),
    secret: process.env.SESSION_SECRET || "super-secret-key",
    resave: false,
    saveUninitialized: false,
    cookie: {
        secure: true,        // Requires HTTPS
        httpOnly: true,      // Prevent JS access
        sameSite: 'strict',  // CSRF protection
        maxAge: 1000 * 60 * 60 // 1 hour
    }
}));

// -----------------------------
// Input Validation Helper
// -----------------------------
function validateInput(username, password) {
    if (!username || !password) return false;
    if (typeof username !== 'string' || typeof password !== 'string') return false;
    if (username.length > 100 || password.length > 200) return false;
    return true;
}

// -----------------------------
// Login Route
// -----------------------------
router.post('/login', loginLimiter, async (req, res) => {
    try {
        const { username, password } = req.body;

        // Validate input
        if (!validateInput(username, password)) {
            return res.status(400).json({
                success: false,
                message: "Invalid credentials"
            });
        }

        // Fetch user from DB (parameterized query)
        const query = "SELECT id, username, password_hash FROM users WHERE username = ?";
        db.get(query, [username], async (err, user) => {
            if (err) {
                console.error(err);
                return res.status(500).json({
                    success: false,
                    message: "Internal server error"
                });
            }

            // Use dummy hash to prevent timing attacks if user not found
            const dummyHash = "$2b$12$C6UzMDM.H6dfI/f/IKcEeOq6W8y8F6LJrVN6a8GN28AL/fOHiceGa";
            const hashToCompare = user ? user.password_hash : dummyHash;

            const passwordMatch = await bcrypt.compare(password, hashToCompare);

            if (!user || !passwordMatch) {
                return res.status(401).json({
                    success: false,
                    message: "Invalid credentials"
                });
            }

            // Regenerate session ID (prevents session fixation)
            req.session.regenerate((err) => {
                if (err) {
                    return res.status(500).json({
                        success: false,
                        message: "Session error"
                    });
                }

                // Store minimal user info in session
                req.session.user = {
                    id: user.id,
                    username: user.username
                };

                return res.status(200).json({
                    success: true,
                    message: "Login successful"
                });
            });
        });

    } catch (error) {
        console.error(error);
        return res.status(500).json({
            success: false,
            message: "Internal server error"
        });
    }
});

module.exports = router;